File: dsa/week(2)/constructor.py
class Student:
    def __init__(self, *args):
        """
        Constructor for the Student class.
        It can be called with 0, 1 or 2 arguments.
        If 0 arguments are provided, it's a default constructor.
        If 1 argument is provided, it's a parameterized constructor with roll number.
        If 2 arguments are provided, it's a parameterized constructor with age and roll number.
        """
        print("Constructor called!")
        if len(args) == 0:
            # Default constructor
            pass
        elif len(args) == 1:
            # Parameterized constructor with roll number
            self.roll_number = args[0]
        elif len(args) == 2:
            # Parameterized constructor with age and roll number
            print("Constructor 3 called!")
            self.age = args[0]
            self.roll_number = args[1]
        else:
            raise TypeError("Invalid number of arguments")

    def display(self):
        try:
            print(f"{self.age} {self.roll_number}")
        except AttributeError:
            print("Age or roll number not set")
    def get_age(self):
        try:
            return self.age
        except AttributeError:
            return None
    def set_age(self, age, password):
        if password != 123:
            return
        if age < 0:
            return
        self.age = age
        
def main():
    
    # Create a student with default constructor
    s3 = Student()
    s3.display()

    s5 = Student(101)
    s5.display()

    # Create a student with age and roll number
    s6 = Student(20, 1002)
    s6.display()


if __name__ == "__main__":
    main()

File: dsa/week(2)/constructor.js
class Student {
    #age;
    constructor(rollNumber = null, age = null) {
        if (rollNumber !== null && age !== null) {
            console.log("this : ", this);
            console.log("Constructor 3 called !");
            this.#age = age;
            this.rollNumber = rollNumber;
        } else if (rollNumber !== null) {
            console.log("Constructor 2 called !");
            this.rollNumber = rollNumber;
        } else {
            console.log("Constructor called !");
        }
    }
    display() {
        console.log(this.#age, this.rollNumber);
    }
    getAge() {
        return this.#age;
    }
    setAge(a, password) {
        if (password !== 123) {
            return;
        }
        if (a < 0) {
            return;
        }
        this.#age = a;
    }
}
function main() {
    let s5 = new Student();
    s5.display();
    console.log("Parameterized constructors Demo");
    let s7 = new Student(101);
    s7.display();
    let s8 = new Student(20, 1002);
    s8.display();
}
main();

